lab/lab-guide.md

```diff
@@ -0,0 +1,802 @@
+    bike products and accessories provided in JSON format below.
+
+    Instructions:
+    - Only answer questions related to the information provided below,
+    - Don't reference any product data not provided below.
+    - If you're unsure of an answer, you can say ""I don't know"" or ""I'm not sure"" and recommend users search themselves.
+
+    Text of relevant information:";
+    ```
+
+Finally, if the responses do not include any information on the bike products being asked, it's possible the call to OpenAI Service is not correct. In the **SemanticKernelService** locate the `GetRagCompletionAsync()` method to make sure that your code matches this sample.
+ 
+    ```csharp
+    public async Task<(string completion, int tokens)> GetRagCompletionAsync(string sessionId, List<Message> contextWindow, string promptText, int productMaxResults)
+    {
+        float[] promptVectors = await GetEmbeddingsAsync(promptText);
+        string productsString = await SearchProductsAsync(promptVectors, productMaxResults);
+
+        var skChatHistory = new ChatHistory();
+        skChatHistory.AddSystemMessage(_systemPromptRetailAssistant + productsString);
+        
+        foreach (var message in contextWindow)
+        {
+            skChatHistory.AddUserMessage(message.Prompt);
+            if (message.Completion != string.Empty)
+                skChatHistory.AddAssistantMessage(message.Completion);
+        }
+
+        PromptExecutionSettings settings = new()
+        {
+            ExtensionData = new Dictionary<string, object>()
+            {
+                { "Temperature", 0.2 },
+                { "TopP", 0.7 },
+                { "MaxTokens", 1000  }
+            }
+        };
+        var result = await kernel.GetRequiredService<IChatCompletionService>().GetChatMessageContentAsync(skChatHistory, settings);
+
+        ChatTokenUsage completionUsage = (ChatTokenUsage)result.Metadata!["Usage"]!;
+        string completion = result.Items[0].ToString()!;
+        int tokens = completionUsage.TotalTokenCount;
+
+        return (completion, tokens);
+    }
+    ```
+</details>
+
+
+# Exercise: Implement a Semantic Cache
+
+Large language models are amazing with their ability to generate completions to a user's questions. However, these requests to generate completions from an LLM are computationally expensive (expressed in tokens) and can also be quite slow. This cost and latency increases as the amount of text increases. 
+
+Applications following the RAG Pattern use data from a database to augment or *ground* the LLM by providing additional information to generate a more contextual response. Payloads now must include the user prompt, conversation history, and additional data from the database, which can get rather large. It is not uncommon to consume thousands of tokens and wait for 3-4 seconds for a response for large payloads. In a world where milliseconds count, waiting 3-4 seconds is often an unacceptable user experience.
```

Comment:
3-4 seconds or even longer!!! I'm finding that most many in here are taking upwards of 10 seconds to complete.